### PR TITLE
Use buildx `docker` driver, add `--cache-to type=inline`

### PR DESCRIPTION
**Pull request by ibeckermayer**
Based on pain experienced attempting to update the `Dockerfile-centos7-fips` image. See the commit messages for details.

Do not merge until: 
- [x] https://github.com/gravitational/teleport.e/pull/3300 is merged and e-ref is updated.
- [x] `Build CentOS 7 Buildbox Assets Images` succeeds, which updates the `build-centos7-assets` image with the new `BUILDKIT_INLINE_CACHE=1` cache: [run](https://github.com/gravitational/teleport/actions/runs/7717269335), [result](https://github.com/gravitational/teleport/pkgs/container/teleport-buildbox-centos7-assets/172883965?tag=teleport16-amd64)
- [x] In the same vain as the prev bullet point, `make -C build.assets/ build-centos7-assets` runs on my arm64 dev machine and the result is manually pushed (we don't have a workflow for this yet): [result](https://github.com/gravitational/teleport/pkgs/container/teleport-buildbox-centos7-assets/172885262?tag=teleport16-arm64)
- [x] ~A `Build CI Buildbox Images` succeeds (TODO): https://github.com/gravitational/teleport/actions/workflows/build-ci-buildbox-images.yaml~: This can't be started manually.
- [x] A dev build succeeds: https://drone.platform.teleport.sh/gravitational/teleport/33501

### CODE REVIEW

**Comment by camscale** [comment]
I think instead of adding a parameter to the CI docker driver setup, we should just add `--builder default` in the `Makefile` where it builds the various buildboxes (6 places). I tested this out and ran a local `make buildbox-centos7-fips`, modified `Dockerfile-centos7` (the base docekrfile) and reran `make buildbox-centos7-fips` and it did the right thing, building a new fips image based on the modified Dockerfile.

Doing this with `--builder default` means it will work from the command line as well as CI.

I don't fully understand the `inline` cache, but I think it should be enabled with `--cache-to type=inline` instead of the buildkit arg as you can only cache to one cache, and this makes it clear you are using that one slot.

But overall it does seem to make sense to use the `docker` (default) builder, We only use the container builder for multi-arch builds of teleport itself. The buildboxes do not have multiarch builds (yet), so I support this change and can see no downside.

**Comment by ibeckermayer** [comment]
> I think instead of adding a parameter to the CI docker driver setup, we should just add `--builder default` in the `Makefile` where it builds the various buildboxes (6 places). I tested this out and ran a local `make buildbox-centos7-fips`, modified `Dockerfile-centos7` (the base docekrfile) and reran `make buildbox-centos7-fips` and it did the right thing, building a new fips image based on the modified Dockerfile.
> 
> Doing this with `--builder default` means it will work from the command line as well as CI.

Agreed that it would be ideal to specify this as a flag, however `default` will not always do the trick. For example in my local setup my builder with the docker driver is named `colima`. For that reason I added a variable to the Makefile in e2aca29d418efb84bee789ac83e351e15489a2d5 for figuring out which builder uses the `docker` driver.

> I don't fully understand the `inline` cache, but I think it should be enabled with `--cache-to type=inline` instead of the buildkit arg as you can only cache to one cache, and this makes it clear you are using that one slot.
> 
> But overall it does seem to make sense to use the `docker` (default) builder, We only use the container builder for multi-arch builds of teleport itself. The buildboxes do not have multiarch builds (yet), so I support this change and can see no downside.

I don't fully understand it either, but basically it means "put the cache alongside the image itself". The alternative options allow for caching to an external repository that's separate from the image. As I understand it, this becomes particularly relevant when doing multi-stage builds wherein you want to cache the intermediate stages without publishing the images themselves.

Confirmation that --cache-to type=inline does the same thing: https://docs.docker.com/build/cache/backends/inline/#synopsis. I changed it to use that since I agree its a better-integrated syntax.

File: build.assets/Makefile [hunk 1]
```diff
@@ -88,6 +88,35 @@ PULL_ON_CI = $(if $(filter $(CI),true),docker inspect --type=image $(1) >/dev/nu
...
+#
+# Third column match:
+#
+# $ docker buildx ls
+# NAME/NODE           DRIVER/ENDPOINT             STATUS   BUILDKIT PLATFORMS
+# container-driver    docker-container
+#   container-driver0 unix:///var/run/docker.sock inactive
+# default *           docker
+#   default           default                     running  20.10.23 linux/arm64, linux/arm/v7, linux/arm/v6
+DOCKER_BUILDER_NAME=$(shell docker buildx ls 2>/dev/null | awk '$$2 == "docker" || $$3 == "docker" {print $$1}')
```
Comment (camscale) [inline build.assets/Makefile:114]: nit: can you put spaces around the `=` - it's a little easier to read and Makefiles are tolerant of spaces there as opposed to shell scripts.

Also, you could hack the field separator in awk like this:

    awk -F '[ *]+' '$$2 == "docker" {print $$1}'

with a comment along the lines of "there may be an asterisk between the first and second column if that record is the default, so just subsume the asterisk into the field separator."

It is a shame that `docker buildx ls` is documented to take a `--format` arg, but does not actually take it (at least on my version of docker) - it would make this cleaner.

File: build.assets/Makefile [hunk 2]
```diff
@@ -88,6 +88,35 @@ PULL_ON_CI = $(if $(filter $(CI),true),docker inspect --type=image $(1) >/dev/nu
 # more details.
 RELEASE_TARGET ?= release
 
+# This variable extracts the name of the Docker buildx builder that uses the Docker driver.
```
Comment (camscale) [inline build.assets/Makefile:91]: Since this line is so far from the variable, perhaps word it as "DOCKER_BUILD_NAME extracts the name of ..." - that also matches Go style.

File: build.assets/Makefile [hunk 3]
```diff
@@ -189,10 +222,21 @@ buildbox-centos7:
 # Builds a Docker buildbox for CentOS 7 FIPS builds
 # Uses regular CentOS 7 buildbox as base, so limited arguments are needed.
 #
+# Because this target uses the image built by the buildbox-centos7 target, it is
+# expected to be run using a docker buildx driver. If another driver is used, the
```
Comment (camscale) [inline build.assets/Makefile:226]: s/a docker/the "docker"/

That currently reads like it just needs some docker buildx driver rather than the specific "docker" buildx driver.

File: build.assets/Makefile [hunk 4]
```diff
@@ -204,12 +248,13 @@ buildbox-centos7-fips: buildbox-centos7
 buildbox-arm:
 	$(call PULL_ON_CI,$(BUILDBOX_ARM))
 	docker buildx build \
-		--build-arg BUILDBOX_VERSION=$(BUILDBOX_VERSION) \
```
Comment (camscale) [inline build.assets/Makefile:207]: Was this meant to be removed? It seems unrelated to the described changes.

**Review by camscale — APPROVED** [review]
Minor non-impacting changes requested, but looks good otherwise. Lets give this a spin - this time it will work for sure!

File: build.assets/Makefile (hunk 1 above)
Comment (ibeckermayer) [reply build.assets/Makefile:114]:
```
awk -F '[ *]+' '$$2 == "docker" {print $$1}'
```

nice one!

> It is a shame that docker buildx ls is documented to take a --format arg, but does not actually take it (at least on my version of docker) - it would make this cleaner.

Yeah, mine neither. Docker is a mess as an overall project.

File: build.assets/Makefile (hunk 4 above)
Comment (ibeckermayer) [reply build.assets/Makefile:207]: Ah yes, this was an opportunistic change since I noticed that Dockerfile-arm doesn't contain any such `ARG`

File: build.assets/Makefile (hunk 3 above)
Comment (ibeckermayer) [reply build.assets/Makefile:208]: 9c41a6810d1eedbf1a9f170ba4c7e8ed6075c288

File: build.assets/Makefile (hunk 2 above)
Comment (ibeckermayer) [reply build.assets/Makefile]: 9c41a6810d1eedbf1a9f170ba4c7e8ed6075c288

File: build.assets/Makefile (hunk 1 above)
Comment (ibeckermayer) [reply build.assets/Makefile]: 9c41a6810d1eedbf1a9f170ba4c7e8ed6075c288

**Comment by ibeckermayer** [comment]
Taking her for another spin: https://drone.platform.teleport.sh/gravitational/teleport/33674

**Comment by ibeckermayer** [comment]
Everything was broken on the last spin for an unrelated reason, here's another after a rebase: https://drone.platform.teleport.sh/gravitational/teleport/33691

File: build.assets/Makefile (hunk 3 above)
Comment (camscale) [reply build.assets/Makefile:208]: This is not address in that commit i think?

Comment (ibeckermayer) [reply build.assets/Makefile:208]: As of that commit it says

> # DOCKER_BUILDER_NAME extracts the name of the buildx builder that uses the "docker" driver.

I think that addresses the original concern, but maybe I didn't understand it properly? Feel free to suggest something else.

Comment (camscale) [reply build.assets/Makefile:208]: That's a different comment - this one says to replace 
> a docker buildx driver

with

> the "docker" buildx driver

Comment (camscale) [reply build.assets/Makefile:208]: that is, in this sentence where my comment is attached, not in the heading.

**Review by jakule — APPROVED** [review]

File: build.assets/Makefile (hunk 3 above)
Comment (ibeckermayer) [reply build.assets/Makefile]: Oh my bad: 75b79ec889e8b02d468ef3bb6d17d37571ed2f9c

**Comment by codingllama** [comment]
FYI, this broke `make enter` for me:

**Comment by codingllama** [comment]
(Edited previous comment with details)
